// ContextMenuManager/index.js
var THREE = require('three');

function ContextMenuManager(app) {
	this.app = app;
	this.showContextMenu = this.showContextMenu.bind(this);
	this.waitingTimeoutId = null;
	this.children = [];
}

var waitDuration = 300;
ContextMenuManager.prototype.showContextMenu = function() {
	var helper = new THREE.AxisHelper(1);
	this.children.push(helper);
	this.app.view.scene.add(helper);
	helper.position.copy(this.targetPos);
	this.waitingTimeoutId = null;
};
ContextMenuManager.prototype.beginWait = function(coord, context) {
	this.optionsContext = context;
	this.targetPos = coord;
	this.waitingTimeoutId = setTimeout(this.showContextMenu, waitDuration);
};
ContextMenuManager.prototype.abortWait = function() {
	if(this.waitingTimeoutId !== null) {
		clearTimeout(this.waitingTimeoutId);
		this.waitingTimeoutId = null;
	}
};
ContextMenuManager.prototype.close = function() {
	this.children.forEach(function(child) {
		child.parent.remove(child);
	});
	this.children.length = 0;
};

module.exports = ContextMenuManager;
